swri_image_util/src/nodelets/replace_colors_nodelet.cpp

```diff
@@ -0,0 +1,457 @@
+    colormap_names_["parula"] = cv::COLORMAP_PARULA;
+    #endif
+  }
+
+  ReplaceColorsNodelet::~ReplaceColorsNodelet()
+  {
+    // Nothing to clean up on destruction
+  }
+
+  // Initialize the nodelet
+  void ReplaceColorsNodelet::onInit()
+  {
+    // Node handles for interacting with ROS
+    ros::NodeHandle& nh = getNodeHandle();
+    ros::NodeHandle& priv_nh  = getPrivateNodeHandle();
+
+    // Lookup table to replace colors with. By default will just convert the
+    // gray scale values to their RGB equivalents. If this node is ever extended
+    // to more than gray scale, this will have to be changed
+    color_lut_ = cv::Mat::zeros(1, NUM_GRAY_VALUES, CV_8UC3);
+    initLut();
+
+    // Get the array representing the color mapping from the parameter server.
+    // This will be in the format [[u0, [r0, g0, b0]], [u1, [r1, g1, b1]]]
+    // where u is the gray scale value to replace, and r, g, b are the RGB
+    // components to replace it with.
+    XmlRpc::XmlRpcValue color_param;
+
+    // This node has two different methods of changing gray scale values to
+    // color imagery. The first maps the gray scale values to OpenCV colormaps
+    // This call checks for that option
+    bool colormap_specified = priv_nh.getParam("colormap", color_param);
+    if (colormap_specified)
+    {
+      //  Use the colormap the user has requested
+      readColormap(color_param);
+    }
+
+    // The other option for modifying the grayscale images is to define a
+    // lookup table that maps grayscale levels to user defined RGB values.
+    // This can be used in conjunction with the colormap option to replace
+    // values in the OpenCV colormap with user values. This call checks
+    // if the user is attempting this operation, and reads the lookup table
+    // from the parameter server if necessary 
+    if (priv_nh.getParam("colors", color_param))
+    {
+      // Try to read in the lookup values from the parameter server.
+      readUserLut(color_param);
+    }
+    else if (!colormap_specified)
+    {
+      ROS_ERROR("Color transformation was not specified. Images will ");
+      ROS_ERROR("only be converted to their gray scale equivalents");
+    }
+
+    // Set up the ROS interface
+    image_transport::ImageTransport it(nh);
+    image_pub_ = it.advertise("modified_image", 1);
+    image_sub_ = it.subscribe(
+      "image",
+      1,
+      &ReplaceColorsNodelet::imageCallback,
+      this);
+  }
+
+  // Callback for getting the input image to change the colors on
+  void ReplaceColorsNodelet::imageCallback(
+    const sensor_msgs::ImageConstPtr& image_msg)
+  {
+    // Only do the color conversion if someone is subscribing to the data
+    if (image_pub_.getNumSubscribers() == 0)
+    {
+      return;
+    }
+
+    // This node currently only support changing gray scale images
+    if (image_msg->encoding != sensor_msgs::image_encodings::MONO8)
+    {
+      ROS_ERROR("Changing image colors is only supported for MONO8 images");
```

Comment:
If an image encoding is incorrect, this will possibly be called several times per second. `ROS_ERROR_THROTTLE` would keep this from flooding the console.